### ISSUE TITLE
Experimental Hyperparameters of IPC=200 on ImageNet-1k

### ISSUE THREAD

**Issue by CYDaaa30** [VILA-Lab/SRe2L#27]
Thank you for your outstanding work! I followed the hyperparameter settings from the CDA method's appendix for the ImageNet-1k experiment to reproduce the validation results at ipc=200, but I obtained a test accuracy of 62.7%, which still has a slight gap compared to the original result of 63.3%. Below are the hyperparameters I used. For hyperparameters not listed in the CDA paper (e.g., temperature, accumulation steps...), I followed the default settings from SRe2L. Could you please check if there might be any issues with these settings?

Optimizer: AdamW
Base learning rate: 1e-3
Weight decay: 1e-2
Batch size: 128
Learning rate schedule: Cosine decay
Training epochs: 300
Augmentation: RandomResizedCrop
Temperature: 20
Gradient accumulation steps: 1
mix_type: cutmix
cutmix: 1.0

With this set of parameters, my reproduction results match the original for ipc=50 and 100, but I encountered this issue with ipc=200.

Additionally, I used the distilled data you provided on Huggingface (cda_in1k_rn18_4k_ipc200), so the distillation quality should not be a factor here.

Thank you for your time and any insights you might have!

**Comment by CYDaaa30** [comment]
> Thank you for your outstanding work! I followed the hyperparameter settings from the CDA method's appendix for the ImageNet-1k experiment to reproduce the validation results at ipc=200, but I obtained a test accuracy of 62.7%, which still has a slight gap compared to the original result of 63.3%. Below are the hyperparameters I used. For hyperparameters not listed in the CDA paper (e.g., temperature, accumulation steps...), I followed the default settings from SRe2L. Could you please check if there might be any issues with these settings?
> 
> Optimizer: AdamW Base learning rate: 1e-3 Weight decay: 1e-2 Batch size: 128 Learning rate schedule: Cosine decay Training epochs: 300 Augmentation: RandomResizedCrop Temperature: 20 Gradient accumulation steps: 1 mix_type: cutmix cutmix: 1.0
> 
> With this set of parameters, my reproduction results match the original for ipc=50 and 100, but I encountered this issue with ipc=200.
> 
> Additionally, I used the distilled data you provided on Huggingface (cda_in1k_rn18_4k_ipc200), so the distillation quality should not be a factor here.
> 
> Thank you for your time and any insights you might have!

**Comment by zeyuanyin** [comment]
@CYDaaa30 Thanks for your attention to our project and your detailed feedback.

I checked my experiment settings on the `cda_in1k_rn18_4k_ipc200`. 

The setting you used is correct, and I posted the [output.log](https://github.com/user-attachments/files/19040119/output.log) here for your reference.


The gap may come from two reasons:

1. Variance in each run.

2. Different dataloader: The only difference between my used code and the published [train_KD.py](https://github.com/VILA-Lab/SRe2L/blob/main/SRe2L/validate/train_KD.py) is the dataloader. I used the `nvidia.dali.pipeline`  to speed up the data loading, which is not included in this repo. But I don't think it is the main reason because I double-checked the output of the DALI loader and Torch loader, and the difference is very slight and almost negligible.

**Comment by zeyuanyin** [comment]
If you want to try  `nvidia.dali.pipeline`, you can follow the code below


```
### utils_dali.py
try:
    from nvidia.dali.plugin.pytorch import DALIClassificationIterator, LastBatchPolicy
    from nvidia.dali.pipeline import pipeline_def
    import nvidia.dali.types as types
    import nvidia.dali.fn as fn
except ImportError:
    raise ImportError("Please install DALI from https://www.github.com/NVIDIA/DALI to run this example.")

@pipeline_def
def create_dali_pipeline(data_dir, crop, size, shard_id, num_shards, dali_cpu=False, is_training=True):
    images, labels = fn.readers.file(file_root=data_dir,
                                     shard_id=shard_id,
                                     num_shards=num_shards,
                                     random_shuffle=is_training,
                                     pad_last_batch=True,
                                     name="Reader")
    dali_device = 'cpu' if dali_cpu else 'gpu'
    decoder_device = 'cpu' if dali_cpu else 'mixed'
    # ask nvJPEG to preallocate memory for the biggest sample in ImageNet for CPU and GPU to avoid reallocations in runtime
    device_memory_padding = 211025920 if decoder_device == 'mixed' else 0
    host_memory_padding = 140544512 if decoder_device == 'mixed' else 0
    # ask HW NVJPEG to allocate memory ahead for the biggest image in the data set to avoid reallocations in runtime
    preallocate_width_hint = 5980 if decoder_device == 'mixed' else 0
    preallocate_height_hint = 6430 if decoder_device == 'mixed' else 0
    if is_training:
        images = fn.decoders.image_random_crop(images,
                                               device=decoder_device, output_type=types.RGB,
                                               device_memory_padding=device_memory_padding,
                                               host_memory_padding=host_memory_padding,
                                               preallocate_width_hint=preallocate_width_hint,
                                               preallocate_height_hint=preallocate_height_hint,
                                            #    random_aspect_ratio=[0.8, 1.25],
                                            #    random_area=[0.1, 1.0],
                                                random_aspect_ratio=[3 / 4, 4 / 3],
                                               random_area=[0.08, 1.0],
                                               num_attempts=100)
        images = fn.resize(images,
                           device=dali_device,
                           resize_x=crop,
                           resize_y=crop,
                           interp_type=types.INTERP_TRIANGULAR)
        mirror = fn.random.coin_flip(probability=0.5)
    else:
        images = fn.decoders.image(images,
                                   device=decoder_device,
                                   output_type=types.RGB)
        images = fn.resize(images,
                           device=dali_device,
                           size=size,
                           mode="not_smaller",
                           interp_type=types.INTERP_TRIANGULAR)
        mirror = False

    images = fn.crop_mirror_normalize(images.gpu(),
                                      dtype=types.FLOAT,
                                      output_layout="CHW",
                                      crop=(crop, crop),
                                      mean=[0.485 * 255,0.456 * 255,0.406 * 255],
                                      std=[0.229 * 255,0.224 * 255,0.225 * 255],
                                      mirror=mirror)
    labels = labels.gpu()
    return images, labels
```


```
### add into train_KD.py
from utils_dali import create_dali_pipeline, DALIClassificationIterator, LastBatchPolicy


    args.local_rank=0
    args.world_size=1
    train_pipe = create_dali_pipeline(batch_size=args.batch_size,
                                        num_threads=args.workers,
                                        device_id=args.local_rank,
                                        data_dir=args.train_dir,
                                        crop=224,
                                        size=256,
                                        shard_id=args.local_rank,
                                        num_shards=args.world_size,
                                        is_training=True)
    train_pipe.build()
    train_loader = DALIClassificationIterator(train_pipe, reader_name="Reader",
                                                last_batch_policy=LastBatchPolicy.PARTIAL,
                                                auto_reset=True)

    val_pipe = create_dali_pipeline(batch_size=args.batch_size,
                                num_threads=args.workers,
                                device_id=args.local_rank,
                                data_dir=args.val_dir,
                                crop=224,
                                size=256,
                                shard_id=args.local_rank,
                                num_shards=args.world_size,
                                is_training=False)
    val_pipe.build()
    val_loader = DALIClassificationIterator(val_pipe, reader_name="Reader",
                                            last_batch_policy=LastBatchPolicy.PARTIAL,
                                            auto_reset=True)
```
